feat(router, docs): Improve authentication guide (#24185)
expo/expo: docs/pages/router/reference/authentication.mdx

title: Authentication in Expo Router
sidebar_title: Authentication
description: How to protect routes with Expo Router.
---

import { Lock01Icon, LockUnlocked01Icon } from '@expo/styleguide-icons';
import { Collapsible } from '~/ui/components/Collapsible';
import { FileTree } from '~/ui/components/FileTree';
import { Step } from '~/ui/components/Step';

> With Expo Router, all routes are always defined and accessible. You can use runtime logic to redirect users away from specific screens based on **authorization**. There are two different techniques for authenticating users within routes. This guide provides an example that demonstrates the functionality of standard native apps.

Reviewer: I will update this in a follow up PR. Thank you for the input!

```
Error: Attempted to navigate before mounting the Root Layout component. Ensure the Root Layout component is rendering a Slot, or other navigator on the first render.
```

To fix this, add a group and move conditional logic down a level.

### Before

<FileTree files={['app/_layout.js', 'app/about.js']} />

```js app/_layout.js
export default function RootLayout() {
  React.useEffect(() => {
    // This navigation event will trigger the error above.
    router.push('/about');
  }, []);

  // This conditional statement creates a problem since the root layout's
  // content (the Slot) must be mounted before any navigation events occur.
  if (isLoading) {
    return <Text>Loading...</Text>;
  }

  return <Slot />;
}
```

### After

<FileTree
  files={[
    ['app/_layout.js', ''],
    ['app/(app)/_layout.js', 'Move conditional logic down a level'],
    'app/(app)/about.js',
  ]}
/>

```js app/_layout.js
export default function RootLayout() {
  return <Slot />;
}
```

```js app/(app)/_layout.js
export default function RootLayout() {
  React.useEffect(() => {
    // This navigation event will trigger the error above.
    router.push('/about');

>   }, []);

Reviewer 2: Agreed, this comment seems like a typo.

Also, the use case is still unclear -- what if the root layout is the one that needs to handle the redirect logic?

>   }, []);

Reviewer: @buraks, it is a typo. Will fix it.

import * as SecureStore from 'expo-secure-store';
import * as React from 'react';
import { Platform } from 'react-native';

type UseStateHook<T> = [[boolean, T | null], (value?: T | null) => void];

function useAsyncState<T>(initialValue?: [boolean, T | null]): UseStateHook<T> {
  return React.useReducer(
    (state: [boolean, T | null], action?: T | null) => [
      false,
      action === undefined ? null : action,
    ],
    initialValue ?? [true, undefined]
  ) as UseStateHook<T>;
}

export async function setStorageItemAsync(key: string, value: string | null) {
  if (Platform.OS === 'web') {
    if (value == null) {
      localStorage.removeItem(key);
    } else {
      localStorage.setItem(key, value);
    }
  } else {
    if (value == null) {
      await SecureStore.deleteItemAsync(key);
    } else {
      await SecureStore.setItemAsync(key, value);
    }
  }
}

export function useStorageState(key: string): UseStateHook<string> {
  // Public
  const [state, setState] = useAsyncState<string>();

  // Get
  React.useEffect(() => {
    if (Platform.OS === 'web') {
      if (typeof localStorage !== 'undefined') {
        setState(localStorage.getItem(key));

>       }

Author: `typeof localStorage` ensures we don't invoke it in Node.js (even though it's already in a `useEffect`).

export default function Root() {
  // Setup the auth context and render our layout inside of it.
  return (
    <SessionProvider>
      <Slot />
    </SessionProvider>
  );
}
```

Next, we'll create a nested [Layout Route](/routing/layouts) which will be used to protect all of the children. This layout will be used to redirect users to the sign in screen if they are not authenticated.

```jsx app/(app)/_layout.js
import { Link, Redirect, Stack } from 'expo-router';
import { Text, View } from 'react-native';

import { useSession } from '../../ctx';

export default function AppLayout() {
  const { session, isLoading } = useSession();

  // You can keep the splash screen open, or render a loading screen.
  if (isLoading) {
    return <Text>Loading...</Text>;
  }

  // Only do authorization on the (app) group as users should always be able to access the (auth) group and sign-in again.
  if (!session) {
    // On web, static rendering will stop here as the user is not authenticated in the headless Node process that the pages are rendered in.
    return <Redirect href="/sign-in" />;
  }

  // This layout can be deferred because it's not the root layout.
  return <Stack />;
}
```

> Now we can create our `/sign-in` page, which can toggle the authentication using `signIn()`. Because this screen is outside of the `(app)` group, it won't be protected by the layout.

Reviewer 3: ```suggestion
Now we can create our `/sign-in` page, which can toggle the authentication using `signIn()`. Because this screen is outside of the `(app)` group, the `(app)` group's layout and its authentication check are not run when rendering this screen. This lets logged-out users see this screen.
```

```jsx app/sign-in.js
import { router } from 'expo-router';
import { Text, View } from 'react-native';

import { useSession } from '../ctx';

export default function SignIn() {
  const { signIn } = useSession();
  return (
    <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
      <Text
        onPress={() => {
          signIn();
          // Navigate after signing in. You may want to tweak this to ensure sign-in is
          // successful before navigating.
          router.replace('/');
        }}>
        Sign In
      </Text>
    </View>
  );
}
```

> And finally we'll implement an authenticated screen which can sign out.

Reviewer 3: ```suggestion
And finally we'll implement an authenticated screen that lets users sign out.
```

```jsx app/(app)/index.js
import { Text, View } from 'react-native';

import { useSession } from '../../ctx';

export default function Index() {
  const { signOut } = useSession();
  return (
    <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
      <Text
        onPress={() => {
          // The `app/(app)/_layout.tsx` will redirect to the sign-in screen.
          signOut();
        }}>
        Sign Out
      </Text>
    </View>
  );
}
```

> We now have an app that can present a loading state while it checks the initial authentication state, and redirect to the sign-in screen if the user is not authenticated. If a user deep links to any protect routes, they'll be redirect back to the sign-in screen.

Reviewer 3: ```suggestion
We now have an app that can present a loading state while it checks the initial authentication state, and redirect to the sign-in screen if the user is not authenticated. If a user visits a deep link to any routes with the authentication check, they'll be redirected back to the sign-in screen.
```

## Alternative loading states

> In Expo Router, something must be rendered to the screen during the loading of the initial auth state. In the example above, we render a loading state in the app layout. Alternatively, you can make the `index` route a loading state, and move the initial route to something like `/home`, this is akin to how Twitter works.

Reviewer 3: ```suggestion
With Expo Router, something must be rendered to the screen during the loading of the initial auth state. In the example above, we render a loading message in the app layout. Alternatively, you can make the `index` route a loading state, and move the initial route to something like `/home`; this is akin to how Twitter works.
```

## Modals and per-route authentication

> Another common pattern is to render a sign-in modal over the top of the app. This enables you to dismiss partially preserve deep links when the authentication is complete. However, this pattern requires routes to be rendered in the background, these routes will need to handle data loading without authentication.

Reviewer 3: ```suggestion
Another common pattern is to render a sign-in modal over the top of the app. This enables you to dismiss [partially preserve deep links]?? when the authentication is complete. However, this pattern requires routes to be rendered in the background; these routes will need to handle data loading without authentication.
```

<FileTree
  files={[
    ['app/_layout.js', 'Declares global session context'],
    'app/(app)/_layout.js',
    ['app/(app)/sign-in.js', <span>Modally presented over the root</span>],
    ['app/(app)/(root)/_layout.js', <span>Protects child routes</span>],
    [
      'app/(app)/(root)/index.js',
      <span>
        Requires authorization <Lock01Icon className="inline mb-1" />
      </span>,
    ],
  ]}
/>

```jsx app/(app)/_layout.js
import { Stack } from 'expo-router';

export const unstable_settings = {
  initialRouteName: '(root)',
};

export default function AppLayout() {
  return (
    <Stack>
      <Stack name="(root)" />
      <Stack
        name="sign-in"
        options={{
          presentation: 'modal',
        }}
      />
    </Stack>
  );
}
```

## Navigating without navigation

If you encounter the following error:

```
Error: Attempted to navigate before mounting the Root Layout component. Ensure the Root Layout component is rendering a Slot, or other navigator on the first render.
```

> This can occur when the app attempts to perform navigation without a navigator mounted in the [root layout](/router/advanced/root-layout). You can fix this by adding a group and moving conditional logic up a level.

Reviewer 3: ```suggestion
This can occur when the app attempts to perform navigation without a navigator mounted in the [root layout](/router/advanced/root-layout). You can fix this by adding a group and moving conditional logic down a level.
```

Wouldn't this be moving down a level since the logic is now deeper in the file hierarchy and in a nested component?

### Before

<FileTree files={['app/_layout.js', 'app/about.js']} />

```js app/_layout.js
export default function RootLayout() {
  React.useEffect(() => {
    // This navigation event will trigger the error above.
    router.push('/about');
  }, []);

>   // Avoid deferring the root layout, otherwise the navigation event will be triggered before the layout is mounted.

Reviewer 3: ```suggestion
  // This conditional statement creates a problem since the root layout's
  // content (the Slot) must be mounted before any navigation events occur.
```

  if (isLoading) {
    return <Text>Loading...</Text>;
  }

  return <Slot />;
}
```

### After

<FileTree
  files={[
    ['app/_layout.js', ''],
    ['app/(app)/_layout.js', 'Move conditional logic up a level'],
    'app/(app)/about.js',
  ]}
/>

```js app/_layout.js
export default function RootLayout() {
  return <Slot />;
}
```

```js app/(app)/_layout.js
export default function RootLayout() {
  React.useEffect(() => {
    // This navigation event will trigger the error above.
    router.push('/about');
  }, []);

>   // Avoid deferring the root layout, otherwise the navigation event will be triggered before the layout is mounted.

Reviewer 3: ```suggestion
  // It is OK to defer rendering this nested layout's content. We couldn't
  // defer rendering the root layout's content since a navigation event (the
  // redirect) would have been triggered before the root layout's content had
  // been mounted.
```

  if (isLoading) {
    return <Text>Loading...</Text>;
  }

  return <Slot />;
}
```

## Middleware

> Traditionally, websites may leverage some form of server-side redirection to protect routes. As of Expo Router v2, Expo Router on web currently only supports build-time static generation and has no support for custom middleware or serving. This can be added in the future to provide a more optimal web experience, in the meantime, authentication can be implemented by using client-side redirects and a loading state.

Reviewer 3: ```suggestion
Traditionally, websites may leverage some form of server-side redirection to protect routes. As of Expo Router v2, Expo Router on web currently only supports build-time static generation and has no support for custom middleware or serving. This can be added in the future to provide a more optimal web experience and, in the meantime, authentication can be implemented by using client-side redirects and a loading state.
```
